File: src/main/java/com/vaadin/demo/component/combobox/ComboBoxConstraints.java
package com.vaadin.demo.component.combobox;

import com.vaadin.flow.component.combobox.ComboBox;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("combo-box-constraints")
public class ComboBoxConstraints extends HorizontalLayout {

    public ComboBoxConstraints() {
        setPadding(false);



        ////  CHECK IF PATTERN ACTUALLY WORKS IN CB!!!

        // tag::snippet[]
        ComboBox<String> field = new ComboBox<>();
        field.setRequired(true);
        field.setAllowedCharPattern("[A-Z]");
        // end::snippet[]
        field.setLabel("Country code");
        field.setHelperText("2-letter uppercase ISO country code");
        field.setItems("DE", "FI", "US");
        field.setAllowCustomValue(true);

        add(field);
    }

    public static class Exporter extends DemoExporter<ComboBoxConstraints> { // hidden-source-line
    } // hidden-source-line
}

File: frontend/demo/component/combobox/combo-box-constraints.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import '@vaadin/combo-box';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('combo-box-constraints')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  protected override render() {
    return html`
        <!-- tag::snippet[] -->
        <vaadin-combo-box
          required
          allowed-char-pattern="[A-Z]"
          label="Country code"
          helper-text="2-letter uppercase ISO country code"
          allow-custom-value
          .items="${['DE','FI','US']}">
        </vaadin-combo-box>
        <!-- end::snippet[] -->
    `;
  }

}
